File: src/NodeParser/NeverTypeNodeParser.ts
import ts from "typescript";
import { Context } from "../NodeParser.js";
import { SubNodeParser } from "../SubNodeParser.js";
import { BaseType } from "../Type/BaseType.js";
import { NeverType } from "../Type/NeverType.js";

export class NeverTypeNodeParser implements SubNodeParser {
    public supportsNode(node: ts.KeywordTypeNode): boolean {
        return node.kind === ts.SyntaxKind.NeverKeyword;
    }
    public createType(_node: ts.KeywordTypeNode, _context: Context): BaseType {
        return new NeverType();
    }
}

File: internal/parser/never_type_node_parser.go
package parser

import (
	"github.com/microsoft/typescript-go/shim/ast"

	"github.com/vega/ts-json-schema-generator/internal/types"
)

// NeverTypeNodeParser handles the `never` keyword
// (src/NodeParser/NeverTypeNodeParser.ts).
type NeverTypeNodeParser struct{}

func NewNeverTypeNodeParser() *NeverTypeNodeParser {
	return &NeverTypeNodeParser{}
}

func (p *NeverTypeNodeParser) SupportsNode(node *ast.Node) bool {
	return node.Kind == ast.KindNeverKeyword
}

func (p *NeverTypeNodeParser) CreateType(_ *ast.Node, _ *Context, _ *types.ReferenceType) types.Type {
	return &types.NeverType{}
}
